### ISSUE TITLE
When I use oci private repository in helm on  kustomization It has a issue.

### ISSUE THREAD

**Issue by huntedhappy** [argoproj/argo-cd#16623]
Checklist:

* [ ] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ ] I've included steps to reproduce the bug.
* [ ] I've pasted the output of `argocd version`.

**Describe the bug**

When I set up kustomization.yaml in my lab It was going well and so I used argo with git but it is not work. I don't know this issue is bug or not.

I set up kustomization.yaml is like below.

It hasn't been works

vi kustomization.yaml

helmCharts:
  - name: nginx
    repo: oci://harbor-infra.huntedhappy.kro.kr/helm
    version: 15.4.4
    releaseName: nginx
    namespace: nginx
    valuesFile: values.yaml
    includeCRDs: false

**To Reproduce**

**Expected behavior**

If I use public repo on helm It works.
for instanace.

vi kustomization.yaml
helmCharts:
  - name: nginx
    repo: https://charts.bitnami.com/bitnami
    version: 15.4.4
    releaseName: nginx
    namespace: nginx
    valuesFile: values.yaml
    includeCRDs: false

**Screenshots**
![argo](https://github.com/argoproj/argo-cd/assets/62123205/7f31a44b-34c0-43a8-b7a4-376202e9ade3)

**Version**

```shell
argocd: v2.9.3+6eba5be
  BuildDate: 2023-12-01T23:24:09Z
  GitCommit: 6eba5be864b7e031871ed7698f5233336dfe75c7
  GitTreeState: clean
  GoVersion: go1.21.4
  Compiler: gc
  Platform: linux/amd64
```

**Logs**

```
Paste any relevant application logs here.Failed to load target state: failed to generate manifest for source 1 of 1: rpc error: code = Unknown desc = Manifest generation error (cached): plugin sidecar failed. error generating manifests in cmp: rpc error: code = Unknown desc = error generating manifests: `sh -c "kustomize build --enable-helm"` failed exit status 1: Error: Error: looks like "oci://harbor-infra.huntedhappy.kro.kr/helm" is not a valid chart repository or cannot be reached: object required : unable to run: 'helm pull --untar --untardir /tmp/_cmp_server/9269bcaa-94c3-4bd2-b64a-fb6fcc7d92d4/nginx/charts --repo oci://harbor-infra.huntedhappy.kro.kr/helm nginx --version 15.4.4' with env=[HELM_CONFIG_HOME=/tmp/kustomize-helm-3756408681/helm HELM_CACHE_HOME=/tmp/kustomize-helm-3756408681/helm/.cache HELM_DATA_HOME=/tmp/kustomize-helm-3756408681/helm/.data] (is 'helm' installed?): exit status 1
```

**Comment by rasheedamir** [comment]
Facing the same issue

**Comment by rasheedamir** [comment]
This is nothing to do with argo-cd! It's a kustomize limitation - https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/helmcharts/#long-term-support

**Comment by fandujar** [comment]
@huntedhappy @rasheedamir I'm working on a PR to workaround it. It will add a helm registry login on the kustomize manifest inflate.

I talked to @crenshaw-dev about it.

I don't know if it will cover all cases but will certainly work for private helm repositories on Google Cloud.

**Comment by paulSambolin** [comment]
@rasheedamir kustomize does support support private OCI registries, it is new and the docs are out of date.  For example, we are using ECR as OCI.  Locally login `aws ecr get-login-password | helm registry login --username AWS --password-stdin 111111111111.dkr.ecr.us-east-1.amazonaws.com` and then pull the helm chart and render the manifests `kustomize build --enable-helm`.  Locally it uses the credentials stored in `/Users/paulsambolin/Library/Preferences/helm/registry/config.json`.

@cbui got it working with argo too.  First edit the `argocd-repo-server` deployment to set `readOnlyRootFilesystem: false`.  Then exec to the `argocd-repo-server` pod and create `~/.docker/config.json`.  This was just a test, we will be injecting credentials using a k8s secret and volume mount.  Because we use ECR, the credentials expire after 24 hours and we will be automatically refreshing the secret

~/.docker/config.json on `argocd-repo-server`
```json
{
        "auths": {
                "111111111111.dkr.ecr.us-east-1.amazonaws.com": {
                        "username": "AWS",
                        "password": "...."
                }
        }
}
```

**Comment by marcusnh** [comment]
I'm experiencing a similar issue, but the problem seems to be with authentication even though we have created a secret to authenticate towards the ACR. See my [Github issue](https://github.com/argoproj/argo-cd/issues/16894) for more information

**Comment by zimmertr** [comment]
> @rasheedamir kustomize does support support private OCI registries, it is new and the docs are out of date. For example, we are using ECR as OCI. Locally login `aws ecr get-login-password | helm registry login --username AWS --password-stdin 111111111111.dkr.ecr.us-east-1.amazonaws.com` and then pull the helm chart and render the manifests `kustomize build --enable-helm`. Locally it uses the credentials stored in `/Users/paulsambolin/Library/Preferences/helm/registry/config.json`.
> 
> @cbui got it working with argo too. First edit the `argocd-repo-server` deployment to set `readOnlyRootFilesystem: false`. Then exec to the `argocd-repo-server` pod and create `~/.docker/config.json`. This was just a test, we will be injecting credentials using a k8s secret and volume mount. Because we use ECR, the credentials expire after 24 hours and we will be automatically refreshing the secret
> 
> ~/.docker/config.json on `argocd-repo-server`
> 
> ```json
> {
>         "auths": {
>                 "111111111111.dkr.ecr.us-east-1.amazonaws.com": {
>                         "username": "AWS",
>                         "password": "...."
>                 }
>         }
> }
> ```

<hr>

Can you give a redacted example of your Kustomization file? Mine looks like this:

```yaml
helmCharts:
  - name: example-chart
    repo: oci://REDACTED.dkr.ecr.us-west-2.amazonaws.com/example-chart
    version: 0.2.0
    namespace: example-app
    releaseName: example-service
    valuesFile: values.yml
```

Which produces this command (shown in STDERR) when running `kustomize build`:

> error: Error: looks like "oci://REDACTED.dkr.ecr.us-west-2.amazonaws.com/example-chart" is not a valid chart repository or cannot be reached: object required
: unable to run: '**helm pull --untar --untardir /REDACTED/charts --repo oci://REDACTED.dkr.ecr.us-west-2.amazonaws.com/example-chart example-chart --version 0.2.0**' with env=[HELM_CONFIG_HOME=/var/folders/_n/1n3wp4qj1v38xjf_32bnpp2m0000gn/T/kustomize-helm-1182230886/helm HELM_CACHE_HOME=/var/folders/_n/1n3wp4qj1v38xjf_32bnpp2m0000gn/T/kustomize-helm-1182230886/helm/.cache HELM_DATA_HOME=/var/folders/_n/1n3wp4qj1v38xjf_32bnpp2m0000gn/T/kustomize-helm-1182230886/helm/.data] (is 'helm' installed?): exit status 1

If I instead run that same command, but remove `--repo`, it works (Although does also return a non-fatal error). But I can't seem to get Kustomize to not pass that argument?

> helm pull --untar --untardir /REDACTED/charts oci://REDACTED.dkr.ecr.us-west-2.amazonaws.com/example-chart example-chart --version 0.2.0
Pulled: REDACTED.dkr.ecr.us-west-2.amazonaws.com/example-chart:0.2.0
Digest: sha256:REDACTED
Error: non-absolute URLs should be in form of repo_name/path_to_chart, got: example-chart

**Comment by zimmertr** [comment]
Nevermind, I think this was introduced in Kustomize 5.2.1 which isn't available in `kubectl` yet. EG:

This doesn't work:
```
$> kubectl version | grep Kustomize
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3

$kubectl kustomize --enable-helm foobar
```

But this does:
```
$> kustomize version
v5.3.0

$> kustomize build --enable-helm foobar
```

**Comment by zimmertr** [comment]
Provided 

> @rasheedamir kustomize does support support private OCI registries, it is new and the docs are out of date. For example, we are using ECR as OCI. Locally login `aws ecr get-login-password | helm registry login --username AWS --password-stdin 111111111111.dkr.ecr.us-east-1.amazonaws.com` and then pull the helm chart and render the manifests `kustomize build --enable-helm`. Locally it uses the credentials stored in `/Users/paulsambolin/Library/Preferences/helm/registry/config.json`.
> 
> @cbui got it working with argo too. First edit the `argocd-repo-server` deployment to set `readOnlyRootFilesystem: false`. Then exec to the `argocd-repo-server` pod and create `~/.docker/config.json`. This was just a test, we will be injecting credentials using a k8s secret and volume mount. Because we use ECR, the credentials expire after 24 hours and we will be automatically refreshing the secret
> 
> ~/.docker/config.json on `argocd-repo-server`
> 
> ```json
> {
>         "auths": {
>                 "111111111111.dkr.ecr.us-east-1.amazonaws.com": {
>                         "username": "AWS",
>                         "password": "...."
>                 }
>         }
> }
> ```

In the case of AWS, one should be able to create a new IAM role with whatever permissions are necessary to read from ECR, and then use IRSA to bind that role to the Repo Server Service Account. Which would mitigate the need to do any filesystem tricks on the pods, no? 

I'll be giving this a try tomorrow... We already use IRSA with Argo to facilitate multi-cluster management. This is a well established pattern for Argo users on AWS.

**Comment by cbui** [comment]
> one should be able to create a new IAM role with whatever permissions are necessary to read from ECR, and then use IRSA to bind that role to the Repo Server Service Account.

I don't think this works.

**Comment by zimmertr** [comment]
@cbui, you're right. Granting the following (All list/read actions to `ecr` and `ecr-public`) to the IRSA-bound role doesn't seem to enable auth/pull. Still getting a 401. Seems like `helm pull` is refusing to use anything but a `config.json` file for auth?

```json
        {
            "Action": [
                "ecr:BatchCheckLayerAvailability",
                "ecr:BatchGetImage",
                "ecr:BatchGetRepositoryScanningConfiguration",
                "ecr:DescribeImageReplicationStatus",
                "ecr:DescribeImageScanFindings",
                "ecr:DescribeRegistry",
                "ecr:DescribeRepositories",
                "ecr:DescribeRepositoryCreationTemplate",
                "ecr:GetAuthorizationToken",
                "ecr:GetDownloadUrlForLayer",
                "ecr:GetLifecyclePolicy",
                "ecr:GetLifecyclePolicyPreview",
                "ecr:GetRegistryPolicy",
                "ecr:GetRegistryScanningConfiguration",
                "ecr:GetRepositoryPolicy",
                "ecr:ListTagsForResource",
                "ecr:ValidatePullThroughCacheRule",
                "ecr:ListImages",
                "ecr:DescribePullThroughCacheRules",
                "ecr:DescribeImages",
                "ecr-public:DescribeImageTags",
                "ecr-public:DescribeRegistries",
                "ecr-public:DescribeRepositories",
                "ecr-public:BatchCheckLayerAvailability",
                "ecr-public:DescribeImages",
                "ecr-public:GetAuthorizationToken",
                "ecr-public:GetRegistryCatalogData",
                "ecr-public:GetRepositoryCatalogData",
                "ecr-public:GetRepositoryPolicy",
                "ecr-public:ListTagsForResource"
            ],
            "Resource": [
                "arn:aws:ecr:$REGION:$ACCOUNT:repository/$REPOSITORY"
            ],
            "Effect": "Allow"
        }
```

What a shame. So the options for now then are (in order of suck):

1. Not use Kustomize to inflate Helm Charts with Argo.
2. Wait for Argo or Kustomize to provide a proper auth mechanism. 
3. Set up some sort of internal proxy to host the chart with internally readable privs.
4. Hack the filesystem of the Repo Server container to include a `config.json` file and periodically rotate the creds.

**Comment by zimmertr** [comment]
What's also interesting, is that our ECR repos are _publicly_ readable internally to the Kubernetes Clusters through Account ID Allow Listing. Which isn't working either.

**Comment by zimmertr** [comment]
Maybe the best way to solve this for now is to use an `initContainer` to auth and create the `config.json` file and save it to a volume and then mount that volume into the Repo Server container. I'm not sure how to handle the cred expiration in that case without killing the repo server pods on a cron though. Meh.

**Comment by zimmertr** [comment]
⚠️ **WARNING** ⚠️  This is a roundabout solution that solves a software problem with infrastructure which is seldom a good idea. If you can afford to, either stop using Kustomize or wait until Argo/Kustomize supports an improved auth model for OCI artifacts in private registries when using Kustomize. 

<hr>

I have a working solution that is automated and doesn't involve any proxies, manual filesystem tweaks, or manual credential rotation. I won't share the exact implementation, but here is the idea:

I deploy Argo CD using the [provided Helm Chart](https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/values.yaml) which exposes the `repoServer.serviceAccount`, `repoServer.extraContainers`, `repoServer.volumes`, and `repoServer.volumeMounts` values. 

Using these values:

1. Define an extraContainer that uses an IAM role with permissions granted to it by IRSA that allows it to run `aws ecr get-login-password` once an hour. Massage the STDOUT from that command into a JSON blob that emulates `config.json`. Save this file to an `emptyDir` volume in the container. I use a script to do this that is volume mounted in from a ConfigMap. 
2. Mount the `emptyDir` volume _again_ into the RepoServer container at a path where RepoServer's Kustomize can find it. Make it ReadOnly on this side. 

The first container will create, and then update, the `config.json` file once an hour and then save it to a volume that is shared with Repo Server. Kustomize, inside the Repo Server container, will then use this config file to auth to ECR and pull the OCI artifact.

**Comment by chancez** [comment]
I hit a bunch of issues trying to get this to work. One issue is that kustomize overrides the `HELM_CONFIG_HOME`, so you can't simply just `helm registry login` anymore. In order to make it work, I found that `HELM_REGISTRY_CONFIG` isn't modified by `kustomize`, so the solution is based on using that to configure where helm looks for registry crews.

Here's my argo-cd helm values.yaml 
```
repoServer:
  initContainers:
    # Setup an init-container to login to private docker registries with 'helm registry login'.
    # This is needed because currently,  the upstream methods of configuring
    # authentication to private helm repositories does not work with
    # kustomize's helmCharts feature and OCI chart repositories.
    - name: registry-auth
      # copied from https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/templates/argocd-repo-server/deployment.yaml
      image: '{{ default .Values.global.image.repository .Values.repoServer.image.repository }}:{{ default (include "argo-cd.defaultTag" .) .Values.repoServer.image.tag }}'
      env:
        - name: HELM_CONFIG_HOME
          value: /helm-auth
        # Configure where helm looks for OCI registry credentials.
        - name: HELM_REGISTRY_CONFIG
          value: /helm-auth/config.json
        - name: QUAY_USERNAME
          valueFrom:
            secretKeyRef:
              name: argocd-quay-helm-credentials
              key: username
        - name: QUAY_PASSWORD
          valueFrom:
            secretKeyRef:
              name: argocd-quay-helm-credentials
              key: password
      volumeMounts:
        # mount the directory that helm will use in the main container
        - mountPath: /helm-auth
          name: registry-auth-dir
      command:
        - /bin/bash
        - -exc
        - 'helm registry login quay.io --username $QUAY_USERNAME --password $QUAY_PASSWORD'
  env:
    - name: HELM_REGISTRY_CONFIG
      value: /helm-auth/config.json
  volumes:
    - name: registry-auth-dir
      emptyDir: {}
  volumeMounts:
    # contains config.json with OCI registry credentials
    - mountPath: /helm-auth
      name: registry-auth-dir
  containerSecurityContext:
    # required for helm to untar the OCI charts
    readOnlyRootFilesystem: true
```

**Comment by TheMatrix97** [comment]
Hi! Any updates on this issue? Thanks!

**Comment by hawkesn** [comment]
This approach worked for me: https://github.com/argoproj/argo-cd/issues/16623#issuecomment-1877669497
Mounting the `~/.docker/config.json` to the repo server.

Thank you @paulSambolin!

This is what worked for me, if you are using the ArgoCD helm chart, in your `values.yaml`:
```
...
repoServer:
  containerSecurityContext:
    readOnlyRootFilesystem: false
  volumes:
    - name: dockerconfigjson
      secret:
        secretName: <name of secret>
        items:
          - key: .dockerconfigjson
            path: config.json
  volumeMounts:
    - name: dockerconfigjson
      mountPath: /home/argocd/.docker/
      readOnly: true
...
```
I am also using External Secrets with the [ECR generator](https://external-secrets.io/v0.9.19/api/generator/ecr/#example-manifest):
Which auto-refreshes the token every 12 hours without needing a script.
```
---
apiVersion: generators.external-secrets.io/v1alpha1
kind: ECRAuthorizationToken
metadata:
  name: ecr-gen
spec:
  region: <your region>
# This pattern above is using the host machines' node iam role to auth
# You will need to adjust this to how you're authenticating, eg. IAM, aws keys, etc.
---
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: <name of external secret>
spec:
  refreshInterval: "1h" # Refresh interval between ES and k8s secret, not re-authing the AWS token
  target:
    name: <name of secret>
    template:
      type: kubernetes.io/dockerconfigjson
      data:
        .dockerconfigjson: |
          {
            "auths": {
              "<AWS_ACCOUNT_ID>.dkr.ecr.<REGION>.amazonaws.com": {
                "username": "AWS",
                "password": "{{ .password }}",
                "auth": "{{ printf "%s:%s" "AWS" .password | b64enc }}"
              }
            }
          }
  dataFrom:
  - sourceRef:
      generatorRef:
        apiVersion: generators.external-secrets.io/v1alpha1
        kind: ECRAuthorizationToken
        name: "ecr-gen"

```
Hope that helps!

**Comment by zimmertr** [comment]
`ECRAuthorizationToken` is cool! Thanks for sharing.

**Comment by kumari-shubham** [comment]
I am also not able to deploy n8n oci chart using argocd, getting the below error 
```
Failed to load target state: failed to generate manifest for source 1 of 1: rpc error: code = Unknown desc = Manifest generation error (cached): `kustomize build <path to cached source>/staging/n8n-cluster --enable-helm` failed exit status 1: Error: Error: could not find protocol handler for: : unable to run: 'helm pull --untar --untardir <path to cached source>/staging/n8n-cluster/charts --repo 8gears.container-registry.com/library/n8n n8n --version 0.23.1' with env=[HELM_CONFIG_HOME=/tmp/kustomize-helm-1677674532/helm HELM_CACHE_HOME=/tmp/kustomize-helm-1677674532/helm/.cache HELM_DATA_HOME=/tmp/kustomize-helm-1677674532/helm/.data] (is 'helm' installed?): exit status 1
```

**Comment by chancez** [comment]
@kumari-shubham you need to use `oci://` in your repository URL

**Comment by bshah0408** [comment]
@chancez  what is a version of Argocd, as well as the version of Kustomize, helm, does this Auth workaround works for OCI://

**Comment by chancez** [comment]
I believe it's kustomize 5.2.1 that supports OCI charts and argocd 2.9.x which updates to kustomize 5.2.1

**Comment by bshah0408** [comment]
@chancez Thank you very much for your reply. we are still behind and I will try to test with Newest version.

**Comment by andrii-korotkov-verkada** [comment]
ArgoCD versions 2.10 and below have reached EOL. Can you upgrade and let us know if the issue is still present, please?

**Comment by jonwinton** [comment]
@andrii-korotkov-verkada I have this issue on `v2.12.4+27d1e64` and this comment still fixed the issue https://github.com/argoproj/argo-cd/issues/16623#issuecomment-2176566772

**Comment by CyDickey-msr** [comment]
I had the same issue but https://github.com/argoproj/argo-cd/issues/16623#issuecomment-2176566772 fixed it as well

**Comment by notniknot** [comment]
In a multi tenancy setup where every project has their own identity token and special permissions for the internal registry, this is still a problem. I'm not aware of any workarounds.

**Comment by cetguyrenard** [comment]
used a variation based on HELM_REGISTRY_CONFIG which seems to work fine

I install argocd via helm and in it's values I put

repoServer:
  env:
    - name: HELM_REGISTRY_CONFIG
      value: /home/argocd/helm/config.json
  volumeMounts:
    - name: oci-ecr-helm-auth
      mountPath: /home/argocd/helm
  volumes:
    - name: oci-ecr-helm-auth
      secret:
        secretName: oci-ecr-helm-auth

where generating ECR secret via external secrets

---
apiVersion: generators.external-secrets.io/v1alpha1
kind: ECRAuthorizationToken
metadata:
  name: ecr-auth-token-gen
  namespace: argocd
spec:
  region: eu-west-1
  auth:
    jwt:
      serviceAccountRef:
        name: argocd-server
---
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: oci-ecr-helm-auth
  namespace: argocd
spec:
  dataFrom:
    - sourceRef:
        generatorRef:
          apiVersion: generators.external-secrets.io/v1alpha1
          kind: ECRAuthorizationToken
          name: ecr-auth-token-gen
  refreshInterval: 30m
  target:
    creationPolicy: Owner
    deletionPolicy: Retain
    name: oci-ecr-helm-auth
    template:
      engineVersion: v2
      mergePolicy: Replace
      data:
        config.json: |
          {
              "auths": {
                      "XXXXXXXXXXX.dkr.ecr.REGION.amazonaws.com": {
                              "username": "{{ .username }}",
                              "password": "{{ .password }}"
                      }
              }
          }

**Comment by reggie-k** [comment]
TLDR:

- The root cause is in Helm support in Kustomize, which does not support private registries:
   https://github.com/kubernetes-sigs/kustomize/issues/4401#issuecomment-1213360696
   From Argo CD side, there is not much that can be done given that statement from Kustomize.
- There are multiple Argo CD configuration workarounds that worked for users for multiple use cases above in this issue.
   More workarounds here: https://github.com/argoproj/argo-cd/issues/10745
- A future action item for Argo CD may be supporting Helm [post-renderers](https://gist.github.com/neoakris/edc0642a088be2cdc4f5ffe8d90ef5ca) as an alternative approach, there is a WIP proposal for this: https://github.com/argoproj/argo-cd/pull/16749

**Comment by israel-morales** [comment]
We were able to resolve this limitation by providing the the `HELM_REGISTRY_CONFIG` environment variable: 

```bash
> kustomize version  
v5.3.0  # Previous versions did not work!

> cat kustomization.yaml 
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

helmCharts:
- name: hello-world
  namespace: hello-world
  releaseName: hello-world
  repo: oci://europe-docker.pkg.dev/our-private-registry/eu.gcr.io/charts
  version: 0.1.0
  valuesFile: values.yaml

> kustomize build --enable-helm
Error: Error: failed to authorize: failed to fetch oauth token: unexpected status from GET request to https://europe-docker.pkg.dev/v2/token?scope=repository.....hello-world%!A(MISSING)pull&service=europe-docker.pkg.dev: 401 Unauthorized

# docker registry credential for the europe-docker.pkg.dev private registry:
> export HELM_REGISTRY_CONFIG=/tmp/config.json   

> kustomize build --enable-helm
apiVersion: v1
kind: Namespace
metadata:
  name: hello-world
...
```

Additionally, if you have argocd plugins with sidecars on the repo-server, they will also need this environment variable. 

Hope this helps

**Comment by notniknot** [comment]
Still, this does not work if you have rbac in your image registry and every user brings their own scoped api key and permissions for only a subset of the images.

**Comment by arch-fan** [comment]
still the same

**Comment by rhoades-brown** [comment]
How can we get https://github.com/argoproj/argo-cd/pull/28678 reviewed by a codeowner?  This PR fixes this issue in an elegant way!
